src/Nethermind/Nethermind.Era1/E2Store.cs

```diff
@@ -0,0 +1,281 @@
+    internal const int ValueSizeLimit = 1024 * 1024 * 50;
+
+    private Stream _stream;
+    private bool _disposedValue;
+
+    private EraMetadata? _metadata;
+    public long StreamLength => _stream.Length;
+    public EraMetadata Metadata => GetMetadata();
+
+    public E2Store(Stream stream)
+    {
+        _stream = stream;
+    }
+
+    private EraMetadata GetMetadata()
+    {
+        if (_metadata is null)
+        {
+            _metadata = ReadEraMetaData().GetAwaiter().GetResult();
+        }
+        return _metadata;
+    }
+
+    public async Task SetMetaData(CancellationToken token = default)
+    {
+        _metadata = await ReadEraMetaData(token);
+    }
+
+    public Task<int> WriteEntryAsSnappy(UInt16 type, byte[] bytes, CancellationToken cancellation = default)
+    {
+        return WriteEntry(type, bytes, true, cancellation);
+
+    }
+    public Task<int> WriteEntry(UInt16 type, byte[] bytes, CancellationToken cancellation = default)
+    {
+        return WriteEntry(type, bytes, false, cancellation);
+    }
+
+    private async Task<int> WriteEntry(UInt16 type, byte[] bytes, bool asSnappy, CancellationToken cancellation = default)
+    {
+        //TODO refactor to sync and use span?
+        byte[] headerBuffer = ArrayPool<byte>.Shared.Rent(HeaderSize);
+        try
+        {
+            //See https://github.com/google/snappy/blob/main/framing_format.txt
+            if (asSnappy && bytes.Length > 0)
+            {
+                //TODO find a way to write directly to file, and still return the number of bytes written
+                using MemoryStream memoryStream = new MemoryStream();
+                using SnappyStream snappyStream = new(memoryStream, CompressionMode.Compress, true);
+                await snappyStream.WriteAsync(bytes, cancellation);
+                await snappyStream.FlushAsync();
+                bytes = memoryStream.ToArray();
+            }
+
+            headerBuffer[0] = (byte)type;
+            headerBuffer[1] = (byte)(type >> 8);
+            int length = bytes.Length;
+            headerBuffer[2] = (byte)(length);
+            headerBuffer[3] = (byte)(length >> 8);
+            headerBuffer[4] = (byte)(length >> 16);
+            headerBuffer[5] = (byte)(length >> 24);
+            headerBuffer[6] = 0;
+            headerBuffer[7] = 0;
+
+            await _stream.WriteAsync(headerBuffer, 0, HeaderSize, cancellation);
+            if (length > 0) await _stream.WriteAsync(bytes, 0, length, cancellation);
+
+            return length + HeaderSize;
+        }
+        finally
+        {
+            ArrayPool<byte>.Shared.Return(headerBuffer);
+        }
+    }
+
+    private async Task<EraMetadata> ReadEraMetaData(CancellationToken token = default)
+    {
+        long l = _stream!.Length;
+        if (_stream.Length < 16)
+            throw new EraFormatException($"Data is not in a valid Era format.");
+
+        byte[] bytes = ArrayPool<byte>.Shared.Rent(16);
```

Comment:
not sure if possible, but consider use `ArrayPoolList` for better readability

Reply:
changed to `ArrayPoolList `